File: Blazor/TodoBlazor/Model/TList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace TodoBlazor.Model
{
	public enum OrderTodo
	{
		None,
		Important,
		EndDate,
		Alphabet,
		CreateDate
	}

	public class TList
	{
		public long? Id { get; set; }
		/// <summary>
		/// Název seznamu
		/// </summary>
		public string Name { get; set; }
		/// <summary>
		/// Jsou otevřené nedokončené úkoly
		/// </summary>
		public bool UnfinishedOpened { get; set; }
		/// <summary>
		/// Jsou zavřené dokončené úkoly
		/// </summary>
		public bool FinishedOpened { get; set; }
		/// <summary>
		/// CSS třída ikony
		/// </summary>
		public string Icon { get; set; }
		/// <summary>
		/// Podle čeho úkoly řadit
		/// </summary>
		public OrderTodo OrderTodo { get; set; }

		public TList()
		{
			GetTodos = async () => { return await Database.Current.GetTodosByParentId(Id.Value); };
			AddTodoD = async (string text) => { await Database.Current.AddTodo(text, Id.Value); };
		}
		public TList(string name, string icon = "", bool unfinishedOpened = true, bool finishedOpened=false, OrderTodo orderTodo = OrderTodo.None)
		{
			Name = name;
			Icon = icon;
			UnfinishedOpened = unfinishedOpened;
			FinishedOpened = finishedOpened;
			OrderTodo = orderTodo;
			GetTodos = async () => { return await Database.Current.GetTodosByParentId(Id.Value); };
			AddTodoD = async (string text) => { await Database.Current.AddTodo(text, Id.Value); } ;
		}

		public TList(string name, Func<Task<List<Todo>>> get, Func<string,Task> add, string icon = "", OrderTodo orderTodo = OrderTodo.None)
		{
			Id = Database.UndividedTListId;
			Name = name;
			Icon = icon;
			UnfinishedOpened = true;
			FinishedOpened = false;
			OrderTodo = orderTodo;
			GetTodos = get;
			AddTodoD = add;
		}

		/// <summary>
		/// Úprava seznamu úkolů v databázi
		/// </summary>
		/// <param name="todoList">Seznam úkolů</param>
		public async Task Update()
		{
			await Database.Current.UpdateTList(this);
		}

		/// <summary>
		/// Smazání seznamu z databáze
		/// </summary>
		/// <param name="id">ID seznamu</param>
		/// <returns></returns>
		public async Task Remove()
		{
			await Database.Current.RemoveTList(Id.Value);
		}

		/// <summary>
		/// Delegát na funkci pro získání úkolů (delegát, aby se tento způsob mohl měnit)
		/// </summary>
		private Func<Task<List<Todo>>> GetTodos;
		/// <summary>
		/// Delegát na funkci pro přidání úkolů (delegát, aby se tento způsob mohl měnit)
		/// </summary>
		private Func<string,Task> AddTodoD;

		public async Task AddTodo(string text)
		{
			await AddTodoD(text);
		}

		public async Task<List<Todo>> GetTodosByOrder()
		{
			List<Todo> todos = await GetTodos();
			return OrderTodo switch
			{
				OrderTodo.None => todos,
				OrderTodo.Important => todos.OrderByDescending(x => x.Important).ToList(),
				OrderTodo.EndDate => todos.OrderBy(x => x.EndDate).ToList(),
				OrderTodo.CreateDate => todos.OrderBy(x => x.CreateDate).ToList(),
				OrderTodo.Alphabet => todos.OrderBy(x => x.Text).ToList(),
				_ => todos,
			};
		}

		/// <summary>
		/// Vrací počet nedokončených úkolů
		/// </summary>
		/// <returns></returns>
		public async Task<string> GetUnfinishedTodosCount()
		{
			int count = (await GetTodos()).Where(x => !x.Done).ToList().Count;
			return count > 0 ? count.ToString() : "";
		}

		public static OrderTodo GetOrderByString(string text)
		{
			if (text == "None")
				return OrderTodo.None;
			else if (text == "Important")
				return OrderTodo.Important;
			else if (text == "CreateDate")
				return OrderTodo.CreateDate;
			else if (text == "EndDate")
				return OrderTodo.EndDate;
			else if (text == "Alphabet")
				return OrderTodo.Alphabet;
			else
				return OrderTodo.None;
		}
	}
}

File: Dart/web/Model/TList.dart
import 'Database.dart';
import 'Todo.dart';

enum eOrderTodo { None, Important, EndDate, Alphabet, CreateDate }

Function DefaultGetTodos = (TList tlist, Function callback) {
  Database.Current.GetTodosByParentId(tlist.Id, callback: (todos) {
    callback(todos);
  });
};

Function DefaultAddTodo = (String text, TList tlist, Function callback) {
  Database.Current.AddTodo(text, tlist.Id, callback: callback);
};

class TList {
  int Id;
  String Name;
  String Icon;
  bool UnfinishedOpened;
  bool FinishedOpened;
  eOrderTodo OrderTodo;
  Function GetTodos;
  Function AddTodo;

  TList(
      {int id = -1,
      String name = '',
      String icon = '',
      bool unfinishedOpened = false,
      bool finishedOpened = false,
      eOrderTodo orderTodo = eOrderTodo.None}) {
    Id = id;
    Name = name;
    Icon = icon;
    UnfinishedOpened = unfinishedOpened;
    FinishedOpened = finishedOpened;
    OrderTodo = orderTodo;
    GetTodos = DefaultGetTodos;
    AddTodo = DefaultAddTodo;
  }

  void Update(callback) {
    //Database.Current().UpdateTList(this, callback);
  }

  void Remove(callback) {
    //Database.Current().RemoveTList(this, callback);
  }

  Map<String, String> toMap() {
    var map = <String, String>{
      'name': Name,
      'icon': Icon,
      'unfinishedOpened': UnfinishedOpened.toString(),
      'finishedOpened': FinishedOpened.toString(),
      'orderTodo': OrderTodo.toString(),
    };
    return map;
  }

  TList.fromMap(Map map, this.Id) {
    Name = map['name'];
    Icon = map['icon'];
    UnfinishedOpened = map['unfinishedOpened'];
    FinishedOpened = map['finishedOpened'];
    OrderTodo = map['orderTodo'];
  }

  void GetTodosByOrder(Function callback) {
    var orderTodo = OrderTodo;
    GetTodos(this, (todos) {
      switch (orderTodo) {
        case eOrderTodo.Important:
          callback(todos.sort((Todo a, Todo b) =>
              (b.Important ? 1 : 0) - (a.Important ? 1 : 0)));
          break;
        case eOrderTodo.EndDate:
          callback(todos.sort((Todo a, Todo b) =>
              DateTime.parse(a.EndDate).compareTo(DateTime.parse(b.EndDate))));
          break;
        case eOrderTodo.Alphabet:
          callback(todos.sort((Todo a, Todo b) =>
              a.Text.toLowerCase().compareTo(b.Text.toLowerCase())));
          break;
        case eOrderTodo.CreateDate:
          callback(todos.sort((Todo a, Todo b) => DateTime.parse(a.CreateDate)
              .compareTo(DateTime.parse(b.CreateDate))));
          break;
        default:
          callback(todos);
          break;
      }
    });
  }

  void GetUnfinishedTodosCount(Function callback) {
    GetTodos(this, (todos) {
      int count = todos.filter((Todo todo) => !todo.Done).length;
      callback(count > 0 ? count : '');
    });
  }
}
